File: example/word-count.py
import bait.zmq

import optparse
import sys

parser = optparse.OptionParser()
parser.add_option("--count", type="int", default=20, help="Number of items to return.  Default: %default.")
parser.add_option("--file2word-workers", type="int", default=2, help="Number of workers.  Default: %default.")
parser.add_option("--count-workers", type="int", default=2, help="Number of workers.  Default: %default.")
(options, arguments) = parser.parse_args()

bait.minnows("filegen", ["localhost"], ["filegen-zmq"] + arguments)
bait.minnows("file2words", ["localhost"] * options.file2word_workers, ["file2words-zmq"])
bait.minnows("count", ["localhost"] * options.count_workers, ["count-zmq"])
bait.minnows("sort", ["localhost"], ["sort-zmq", str(options.count)])
bait.minnows("print", ["localhost"], ["print-zmq"])

bait.hook("filegen", 0, bait.ROUND_ROBIN, 0, "file2words")
bait.hook("file2words", 0, bait.HASHED, 0, "count")
bait.hook("count", 0, bait.SINGLE, 0, "sort")
bait.hook("sort", 0, bait.SINGLE, 0, "print")

bait.start()

File: example/word-count.cpp
#include <phish-bait.hpp>

int main(int argc, char* argv[])
{
  std::vector<std::string> files(argv + 1, argv + argc);
  std::vector<std::string> filegen_hosts(1, "localhost");
  std::vector<std::string> filegen_arguments(1, "filegen-zmq");
  filegen_arguments.insert(filegen_arguments.end(), files.begin(), files.end());
  phish::bait::school("filegen", filegen_hosts, filegen_arguments);

  std::vector<std::string> file2words_hosts(2, "localhost");
  std::vector<std::string> file2words_arguments(1, "file2words-zmq");
  phish::bait::school("file2words", file2words_hosts, file2words_arguments);

  std::vector<std::string> count_hosts(2, "localhost");
  std::vector<std::string> count_arguments(1, "count-zmq");
  phish::bait::school("count", count_hosts, count_arguments);

  std::vector<std::string> sort_hosts(1, "localhost");
  std::vector<std::string> sort_arguments(1, "sort-zmq");
  sort_arguments.insert(sort_arguments.end(), "20"); // Return the top-twenty words
  phish::bait::school("sort", sort_hosts, sort_arguments);

  std::vector<std::string> print_hosts(1, "localhost");
  std::vector<std::string> print_arguments(1, "print-zmq");
  phish::bait::school("print", print_hosts, print_arguments);

  phish::bait::hook("filegen", 0, PHISH_BAIT_ROUND_ROBIN, 0, "file2words");
  phish::bait::hook("file2words", 0, PHISH_BAIT_HASHED, 0, "count");
  phish::bait::hook("count", 0, PHISH_BAIT_SINGLE, 0, "sort");
  phish::bait::hook("sort", 0, PHISH_BAIT_SINGLE, 0, "print");

  phish::bait::start();

  return 0;
}
